flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/AdaptedRestartPipelinedRegionStrategyNGFailoverTest.java

```diff
@@ -0,0 +1,513 @@
+	 *        (blocking)
+	 * </pre>
+	 */
+	@Test
+	public void testFailurePropagationToUnderlyingStrategy() throws Exception {
+		final JobGraph jobGraph = createBatchJobGraph();
+		final ExecutionGraph eg = createExecutionGraph(jobGraph);
+
+		final TestAdaptedRestartPipelinedRegionStrategyNG failoverStrategy =
+			(TestAdaptedRestartPipelinedRegionStrategyNG) eg.getFailoverStrategy();
+
+		final Iterator<ExecutionVertex> vertexIterator = eg.getAllExecutionVertices().iterator();
+		final ExecutionVertex ev11 = vertexIterator.next();
+		final ExecutionVertex ev12 = vertexIterator.next();
+		final ExecutionVertex ev21 = vertexIterator.next();
+
+		// trigger downstream regions to schedule
+		componentMainThreadExecutor.execute(() -> {
+			// finish upstream regions to trigger scheduling of downstream regions
+			ev11.getCurrentExecutionAttempt().markFinished();
+			ev12.getCurrentExecutionAttempt().markFinished();
+		});
+
+		// trigger task failure of ev21 on consuming data from ev11
+		Exception taskFailureCause = new PartitionConnectionException(
+			new ResultPartitionID(
+				ev11.getProducedPartitions().keySet().iterator().next(),
+				ev11.getCurrentExecutionAttempt().getAttemptId()),
+			new Exception("Test failure"));
+		ev21.getCurrentExecutionAttempt().fail(taskFailureCause);
+		manualMainThreadExecutor.triggerAll();
+
+		assertThat(failoverStrategy.getLastTasksToCancel(), containsInAnyOrder(
+			new ExecutionVertexID(ev11.getJobvertexId(), 0),
+			new ExecutionVertexID(ev21.getJobvertexId(), 0),
+			new ExecutionVertexID(ev21.getJobvertexId(), 1)));
+	}
+
+	/**
+	 * Tests that the vertex partition state is reset property during failover.
+	 * <pre>
+	 *     (v11) -+-> (v21)
+	 *            x
+	 *     (v12) -+-> (v22)
+	 *
+	 *            ^
+	 *            |
+	 *        (blocking)
+	 * </pre>
+	 */
+	@Test
+	public void testStatusResettingOnRegionFailover() throws Exception {
+		// create a batch job graph with LAZY_FROM_SOURCES schedule mode
+		final JobGraph jobGraph = createBatchJobGraph();
+		final ExecutionGraph eg = createExecutionGraph(jobGraph);
+		final AdaptedRestartPipelinedRegionStrategyNG strategy =
+			(AdaptedRestartPipelinedRegionStrategyNG) eg.getFailoverStrategy();
+
+		final Iterator<ExecutionVertex> vertexIterator = eg.getAllExecutionVertices().iterator();
+		final ExecutionVertex ev11 = vertexIterator.next();
+		final ExecutionVertex ev12 = vertexIterator.next();
+		final ExecutionVertex ev21 = vertexIterator.next();
+		final ExecutionVertex ev22 = vertexIterator.next();
+
+		// finish upstream regions so that all the blocking result partitions should be finished and consumable
+
+		// finish upstream regions to trigger scheduling of downstream regions
+		ev11.getCurrentExecutionAttempt().markFinished();
+		ev12.getCurrentExecutionAttempt().markFinished();
+
+		// force FINISHED ev11 to fail to reset its partition
+		strategy.onTaskFailure(ev11.getCurrentExecutionAttempt(), new FlinkException("Fail for testing"));
```

Comment:
I don't believe this scenario to be possible in production. FAILED to FINISHED is not a valid state transition. Could we not fail one of the consumers instead?

Reply:
Let's delete the test. The adapter is not calling `IntermediateResultPartition#resetForNewExecution()`